File: c/lab01.c
#include <stdio.h>
#include <stdlib.h>
#include <time.h>
#include <unistd.h>
#include <math.h>
#include "lab01.h"

double * pearson_cor(int ** matrix, int * vector, int n);
double * pearson_cor_o(int ** matrix, int * vector, int n);

int main(int argc, char *argv[]){
  int ** matrix = NULL, * vector = NULL, n = 0;
  double * rho_vector = NULL, time_elapsed;
  clock_t start, end;

  FILE *fptr;

  if (argc < 2) {
    printf("Please enter n (n > 0): ");
    while(n < 1){
      scanf("%d", &n);
      getchar();
    }
  }
 
  if (argc == 2) {
    int n = (int) *argv[1];
  }


  check_if_file_exists(&fptr, "log/results.txt");
  fprintf(fptr, "%d: ", n);
  fclose(fptr);

  srand(time(NULL));

  for(int i = 0; i < 3; i++){
    printf("\n\n[RUN #%d]\n", (i+1));
    printf("\nInitializing matrix...\n");
    start = clock();
    matrix = init_matrix(n, 1, 10);
    end = clock();
    time_elapsed = ((double) end - start) / CLOCKS_PER_SEC;
    if(matrix == NULL){
      return EXIT_FAILURE;
    }
    printf("\nInitialization done. (Completed in %.4f seconds)\n", time_elapsed);

    printf("\nInitializing vector...\n");
    start = clock();
    vector = init_vector(n, 1, 10);
    end = clock();
    time_elapsed = ((double) end - start) / CLOCKS_PER_SEC;
    if(vector == NULL){
      return EXIT_FAILURE;
    }
    printf("\nInitialization done. (Completed in %.4f seconds)\n", time_elapsed);

    printf("\nRunning function pearson_cor()...\n");
    start = clock(); 
    rho_vector = pearson_cor(matrix, vector, n);
    end = clock();
    time_elapsed = ((double) end - start) / CLOCKS_PER_SEC;
    if (rho_vector == NULL){
      printf("Error!\n");
      return EXIT_FAILURE;
    }
    printf("\nElapsed time: %.4f seconds\n", time_elapsed);
    
    
    check_if_file_exists(&fptr, "log/results.txt");
    fprintf(fptr, "%.4f, ", time_elapsed);
    fclose(fptr);

    for (int i = 0; i < n; i++)
      free(matrix[i]);
    free(matrix);
    free(vector);
    free(rho_vector);
    }
 
  check_if_file_exists(&fptr, "log/results.txt");
  fprintf(fptr, "\n");
  fclose(fptr);
  return EXIT_SUCCESS;
}

double * pearson_cor(int ** matrix, int * vector, int n){
  double * rho_vector = NULL;
  int x, x_sq, y, y_sq, xy;
  double numerator, denominator;
  
  rho_vector = (double *) malloc(n * sizeof(double));
  if(vector == NULL)
      return NULL;

  y = y_sq = 0;

  for (int i = 0; i < n; i++){
    y += vector[i];
    y_sq += vector[i] * vector[i];
  }

  for (int i = 0; i < n; i++){
    x = x_sq = xy = 0;

    for (int j = 0; j < n; j++) {
      x += matrix[j][i];
      x_sq += matrix[j][i] * matrix[j][i];
      xy += matrix[j][i] * vector[j];
    }

    numerator = (double) (n * xy) - (y*x);
    denominator = sqrt(((n*x_sq) - (x*x)) * ((n*y_sq) * (y*y)));
    
    rho_vector[i] = (denominator == 0) ? NAN : numerator / denominator;
    track_progress(i+1, n);
  }

  return rho_vector;
}

double * pearson_cor_o(int ** matrix, int * vector, int n){
  double * rho_vector = NULL;
  int x, x_sq, y, y_sq, xy;
  double numerator, denominator;
  
  rho_vector = (double *) malloc(n * sizeof(double));
  if(vector == NULL)
      return NULL;

  y = y_sq = 0;
  for (int i = 0; i < n; i++){
    y += vector[i];
    y += vector[i] * vector[i];
  }

  for (int i = 0; i < n; i++){
    x = x_sq = xy = 0;
    int * vector_col_i = get_vector_from_column(matrix, n, i); //get column as vector
    
    for (int j = 0; j < n; j++) {
      x += vector_col_i[j];
      x_sq += vector_col_i[j] * vector_col_i[j];     
      xy += vector_col_i[j] * vector[j];
    }

    numerator = (double) (n * xy) - (y*x);
    denominator = sqrt(((n*x_sq) - (x*x)) * ((n*y_sq) * (y*y)));
    rho_vector[i] = (denominator == 0) ? NAN : numerator/denominator;
    track_progress(i+1, n);
    free(vector_col_i);
  }

  return rho_vector;
}

File: python/lab01.py
import random
import time


def track_progress(current, total):
    progress = (current / total) * 100.00
    print(f"\rProgress: {progress:.2f}% - {current} out of {total}", end="")


def measure_execution_time(function):
    def wrapper(*args, **kwargs):
        print(f"Starting execution of {function.__name__}...")
        start = time.time()
        result = function(*args, **kwargs)
        end = time.time()
        elapsed_time = end - start
        print(
            f"\rFunction {function.__name__} ran for {elapsed_time:.4f} seconds.")
        return result
    return wrapper


@measure_execution_time
def init_matrix(n, min, max):
    matrix = []

    for i in range(0, n):
        row = []
        for j in range(0, n):
            row.append(random.randint(min, max))
            track_progress(((i * n) + j + 1), n*n)
        matrix.append(row)

    return matrix


@measure_execution_time
def init_vector(n, min, max):
    vector = []

    for i in range(0, n):
        vector.append(random.randint(min, max))
        track_progress(i+1, n)

    return vector


def get_column_i(matrix, i, size):
    column = []

    for iter in range(0, size):
        column.append(matrix[iter][i])

    return column


def print_matrix(n, matrix):
    for i in range(0, n):
        for j in range(0, n):
            print(str(matrix[i][j]), end="\t")
        print("\n")


@measure_execution_time
def pearson_cor(matrix, vector, size):
    rho_vector = []

    for i in range(0, size):
        summ_x = summ_y = summ_x_sq = summ_y_sq = summ_xy = 0
        for j in range(0, size):
            summ_x = summ_x + matrix[j][i]
            summ_y = summ_y + vector[j]
            summ_x_sq = summ_x_sq + (matrix[j][i] ** 2)
            summ_y_sq = summ_y_sq + (vector[j] ** 2)
            summ_xy = summ_xy + (matrix[j][i] * vector[j])
        numerator = (size * summ_xy) - (summ_y * summ_x)
        denominator = ((size * summ_x_sq) - (summ_x ** 2)) * \
            ((size * summ_y_sq) - (summ_y ** 2)) ** 0.5
        rho_vector.append(numerator/denominator)
        track_progress(i+1, size)

    return rho_vector


@measure_execution_time
def pearson_cor_o(matrix, vector, size):
    rho_vector = []

    for i in range(0, size):
        summ_x = summ_y = summ_x_sq = summ_y_sq = summ_xy = 0
        x_col_i = get_column_i(matrix, i, size)
        for j in range(0, size):
            summ_x = summ_x + x_col_i[j]
            summ_y = summ_y + vector[j]
            summ_x_sq = summ_x_sq + (x_col_i[j] ** 2)
            summ_y_sq = summ_y_sq + (vector[j] ** 2)
            summ_xy = summ_xy + (x_col_i[j] * vector[j])
        numerator = (size * summ_xy) - (summ_y * summ_x)
        denominator = ((size * summ_x_sq) - (summ_x ** 2)) * \
            ((size * summ_y_sq) - (summ_y ** 2)) ** 0.5
        rho_vector.append(numerator/denominator)
        track_progress(i+1, size)

    return rho_vector


def main():
    n = 0

    while (n < 1):
        n = int(input("Please enter n: "))

    matrix = init_matrix(n, 1, 10)
    vector = init_vector(n, 1, 10)

    pearson_cor_o(matrix, vector, n)

    matrix = None
    vector = None


if __name__ == "__main__":
    main()
